tests/Doctrine/Tests/Common/Proxy/ProxyMagicMethodsTest.php

```diff
@@ -285,6 +285,24 @@ public function testInheritedMagicSleep()
         self::assertSame('defaultValue', $unserialized->nonSerializedField, 'Field was not returned by "__sleep"');
     }
 
+    public function testInheritedMagicSleepTypehinted()
+    {
+        $proxyClassName = $this->generateProxyClass(MagicSleepClassTypehinted::class);
+        $proxy          = new $proxyClassName();
+
+        $sleepMethod = '__sleep';
+
+        $proxyReflection = new ReflectionClass($proxy);
+        self::assertTrue($proxyReflection->hasMethod($sleepMethod), $sleepMethod . 'method doesn\'t exist');
+
+        $methodReflection = $proxyReflection->getMethod($sleepMethod);
+        self::assertSame(
+            $methodReflection->getReturnType()->getName(),
```

Comment:
What guarantees that this will return a `ReflectionNamedType`, not just a `ReflectionType`?

Reply:
`getReturnType` will return a `ReflectionNamedType` in the supported PHP versions, unless we're on PHP 8 and the method has a union return type, in which case this is a `ReflectionUnionType`. So there should be a conditional to support both, as `ReflectionUnionType` doesn't have a `getName` method.

Reply:
Should we check something in case of `ReflectionUnionType`?
In context of `__sleep` and `__wakeup` there not so much variability, I suppose.

Reply:
[Here](https://wiki.php.net/rfc/union_types_v2#reflection) I see `__toString` method but it confuses me that it is deprecated in parent `ReflectionType` class.

Reply:
Yes, the method is deprecated. For `ReflectionNamedType` (which is always returned for `getReturnType` starting with PHP 7.1.0), you should use the `getName()` method. Not sure what the alternative is for `ReflectionUnionType`.